### PR TITLE
[1943] Add Settings::validate to reject conflicting configuration

### LINKED ISSUE #1943: Add `Settings::validate`
**Is your feature request related to a problem? Please describe.**
There are many configuration options in Settings. Some of them conflict with each other. For example:

https://github.com/slatedb/slatedb/pull/1939/changes#r3597809029

**Describe the solution you'd like**
I'd like SlateDB to validate that settings fields are at least compatible with each other and don't lead to any obvious deadlocks or failures.

**Describe alternatives you've considered**
Not doing this. That's what we do today. Many of the configurations are unlikely unless the user is being adversarial or exploring random state (as we do in DST).

**Additional context**
I believe this was discussed during the LSM segment tree work as well, but I can't find the PR conversation.

### PR DESCRIPTION
**Pull request by 1996fanrui**
## Summary

Closes #1943.

`Settings` has many fields that are individually valid but conflict when
combined, leading to runtime deadlocks or failures. The motivating case (from
[#1939 review](https://github.com/slatedb/slatedb/pull/1939/changes#r3597809029)):
when `max_unflushed_bytes <= l0_sst_size_bytes`, a single write can push the
active memtable past the backpressure limit before it is ever frozen. With no
immutable memtable or WAL to drain, `maybe_apply_backpressure` blocks the next
write forever and spins on CPU — for an otherwise "valid" configuration.

This adds a `Settings::validate` that rejects such combinations up front so
misconfiguration fails loudly when the database is opened rather than hanging
later.

## Changes

- Add `Settings::validate`, which centralizes the two configuration checks
  previously inlined in `DbBuilder::build` (`l0_flush_parallelism >= 1`,
  `max_wal_flushes_before_l0_flush >= 4096`) and adds the new
  `max_unflushed_bytes > l0_sst_size_bytes` rule.
- Call `self.settings.validate()?` at the top of `DbBuilder::build` and remove
  the inlined checks, so all construction paths validate through one entry point.
- In the DST `build_settings`, derive `max_unflushed_bytes` from
  `l0_sst_size_bytes` (lower bound `l0_sst_size_bytes + 1`) so randomized
  configs never generate the invalid combination — matching the existing
  pattern used for `worker_heartbeat_timeout` and `checkpoint_lifetime`.
- Add unit tests covering the valid default, the strict-greater boundary, and
  each rejection case.

## Notes for Reviewers

The two migrated checks keep identical error messages and behavior; the existing
`DbBuilder` tests that assert those rejections pass unchanged. Cross-field
validation of `CompactorOptions` internals (heartbeat timeout vs. interval,
min/max compaction sources) is intentionally out of scope here — those live
behind a `HashMap<String, String>` scheduler config and can be added
incrementally as the issue suggests.

## Checklist

- [x] Small, scoped PR (< 500 total lines excluding tests); or opened as Draft with a plan on how to break it into smaller pieces
- [x] Linked related issue(s) or added context in the description
- [x] Self-reviewed the diff; added comments for tricky parts
- [x] Tests added/updated and passing locally
- [x] Ran `cargo fmt`, `cargo clippy --all-targets --all-features`, and `cargo nextest run --all-features`
- [x] Called out any breaking changes and provided migration notes
- [x] Considered performance impact; added notes or benchmarks if relevant

Thank you for the review! 🙏

### CODE REVIEW

File: slatedb/src/config.rs [hunk 1]
```diff
@@ -820,6 +820,37 @@ impl Settings {
...
     }
 
+    /// Validates that the settings are internally consistent, rejecting field
+    /// combinations that would deadlock or fail at runtime.
+    ///
+    /// # Errors
+    ///
+    /// Returns an [`crate::Error`] with [`crate::ErrorKind::Invalid`] describing
+    /// the first invalid setting or combination encountered.
+    pub fn validate(&self) -> Result<(), crate::Error> {
```
Comment (criccomini) [inline slatedb/src/config.rs:830]: Should be using `SlateDBError` here. Probably something like `SlateDBError::InvalidConfig". Then .into() or map_err here

File: slatedb/src/db.rs [hunk 2]
```diff
@@ -5062,12 +5064,15 @@ mod tests {
...
+        // bytes above the backpressure threshold indefinitely.
         let object_store: Arc<dyn ObjectStore> = Arc::new(InMemory::new());
-        let mut options = test_db_options(0, 1024 * 1024, None);
+        let fp_registry = Arc::new(FailPointRegistry::new());
+        fail_parallel::cfg(fp_registry.clone(), "write-compacted-sst-io-error", "pause").unwrap();
+
+        let mut options = test_db_options(0, 4 * 1024, None);
         options.flush_interval = None;
-        options.max_unflushed_bytes = 1;
+        options.max_unflushed_bytes = 8 * 1024;
```
Comment (1996fanrui) [inline slatedb/src/db.rs:5075]: This test previously forced backpressure via `max_unflushed_bytes < l0_sst_size_bytes`, which the new validation now rejects. Reworked to a valid config: pause the L0 upload so a frozen memtable stays in memory, keeping backpressure applied while the DB is fenced. 

The fenced-wakeup behavior under test is unchanged.

File: slatedb/src/config.rs [hunk 3]
```diff
@@ -820,6 +820,37 @@ impl Settings {
...
+        }
+        if self.max_wal_flushes_before_l0_flush < 4096 {
+            return Err(crate::Error::invalid(
+                "invalid configuration: max_wal_flushes_before_l0_flush must be at least 4096"
+                    .into(),
+            ));
+        }
+        // `max_unflushed_bytes` must exceed `l0_sst_size_bytes` so the active
+        // memtable is always frozen before backpressure permanently blocks writes.
+        if self.max_unflushed_bytes <= self.l0_sst_size_bytes {
```
[the remark this answers is not in the document]
Comment (1996fanrui) [reply slatedb/src/config.rs:846]: Good catch — fixed all of these to satisfy the new rule.

File: slatedb/src/config.rs (hunk 1 above)
Comment (1996fanrui) [reply slatedb/src/config.rs:830]: thanks for the quick review, sounds make sense, and `InvalidConfiguration` is introduced.

**Review by criccomini — APPROVED** [review]

**Comment by 1996fanrui** [comment]
@criccomini Thanks for the review, have a nice weekend!
